### PR TITLE
code-substitute: Switch from `block_hash` to `block_number`

### PR DESCRIPTION
**Pull request by bkchr**
This will make it easier for light clients to work with the code-substitute.

For more information on this see: https://github.com/paritytech/substrate/issues/10589

Closes: https://github.com/paritytech/substrate/issues/10589

polkadot companion: https://github.com/paritytech/polkadot/pull/4667

### CODE REVIEW

File: client/service/src/client/wasm_substitutes.rs [hunk 1]
```diff
@@ -41,20 +40,16 @@ struct WasmSubstitute<Block: BlockT> {
 	code: Vec<u8>,
 	hash: Vec<u8>,
 	/// The hash of the block from that on we should use the substitute.
```
Comment (andresilva) [inline client/service/src/client/wasm_substitutes.rs:42]:
```suggestion
	/// The block number on which we should start using the substitute.
```

File: client/service/src/builder.rs [hunk 2]
```diff
@@ -281,14 +279,16 @@ where
...
-			.map(|(h, c)| {
-				let hash = TBl::Hash::from_str(&h).map_err(|_| {
+			.map(|(n, c)| {
+				let number = NumberFor::<TBl>::from_str(&n).map_err(|_| {
 					Error::Application(Box::from(format!(
-						"Failed to parse `{}` as block hash for code substitutes.",
-						h
+						"Failed to parse `{}` as block number for code substitutes. \
+						 In an old version the key for code substitute was a block hash. \
+						 So, please update the chain spec to a version that is compatible with your node.",
```
Comment (andresilva) [inline client/service/src/builder.rs:287]:
```suggestion
						 Please update the chain spec to a version that is compatible with your node.",
```

**Review by andresilva — APPROVED** [review]

**Comment by tomaka** [comment]
This is still not ideal, because if there's a code substitute at block 10, and a light client warp syncs to block 15, it will have to download the runtime at block 10, the runtime at block 15, compare them, and if they're equal then do the substitution.

Basically the problem is that the client has no way to know whether the substitute is still relevant.
If there's a code substitute at block 100 and we warp sync to block 9000000, the light client will still have to download the runtime of block 100 to check whether it is the same.

**Comment by bkchr** [comment]
> This is still not ideal, because if there's a code substitute at block 10, and a light client warp syncs to block 15, it will have to download the runtime at block 10, the runtime at block 15, compare them, and if they're equal then do the substitution.

We already had discussed this once :D You don't need to download the runtime at 10, because you should use the wasm binary provided in the chain spec. You could, as we are doing it, check the spec_version on startup and then you will just need to compare them against the spec_version of 15.



> Basically the problem is that the client has no way to know whether the substitute is still relevant.
> If there's a code substitute at block 100 and we warp sync to block 9000000, even though it's likely that there's been a runtime upgrade in between, the light client has no way to know and will still have to download the runtime of block 100 to check whether it is the same.

Same as above, you have the binary in the chain spec. You don't need to download it, because if you download it you will also not get the correct one. (You will get the one from on chain, while you want the one that is present in the chain spec)

**Comment by tomaka** [comment]
> We already had discussed this once :D You don't need to download the runtime at 10, because you should use the wasm binary provided in the chain spec. You could, as we are doing it, check the spec_version on startup and then you will just need to compare them against the spec_version of 15.

The problem is that the light client can't know whether there has been a runtime upgrade between blocks 10 and 15, or between blocks 100 and 900000.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**